testing/dart/locale_test.dart

```diff
@@ -9,146 +9,261 @@ import 'package:test/test.dart';
+          'interchange.',
+    );
+
+    expect(
+      const Locale('iw').toString(),
+      'he',
+      reason: 'The language code for Hebrew was officially changed in 1989.',
+    );
+    expect(
+      const Locale('iw', 'DD').toString(),
+      'he_DE',
+      reason: 'Legacy. This is a valid Unicode Locale Identifier, even if '
+          'not a valid Unicode BCP47 Locale Identifier',
+    );
+    expect(
+      const Locale('iw', 'DD'),
+      const Locale('he', 'DE'),
+      reason: 'The German Democratic Republic ceased to exist in '
+          'October 1990.',
+    );
   });
 
   test('Locale unnamed constructor idiosyncrasies', () {
-    expect(() => Locale.parse(Locale('en', '').toString()), throwsException,
-        reason: 'Locale("en", "").toString() produces "en-" which is not '
-            'standards-compliant.');
-
-    expect(Locale.parse(Locale('en', '').toLanguageTag()), Locale('en'),
-        reason: 'There is no standards-compliant way for toLanguageTag() to '
-            'represent a zero-length region code.');
-    expect(Locale.parse(Locale('en', '').toLanguageTag()), Locale('en'),
-        reason: 'There is no standards-compliant way for toLanguageTag() to '
-            'represent a zero-length region code.');
-
-    expect(Locale('abcd').toLanguageTag(), 'abcd',
-        reason: 'Locale("abcd") is not following instructions in the API '
-            'documentation, so produces standards-uncompliant output.');
-    expect(Locale.parse('abcd').toLanguageTag(), 'und-Abcd', reason: '');
-    expect(Locale('abcd'), isNot(Locale.parse('abcd')), reason: '');
-    expect(() => Locale.parse(Locale('a').toLanguageTag()), throwsException,
-        reason: 'Locale("abcd") is not following instructions in the API '
-            'documentation, so produces standards-uncompliant output.');
-    expect(Locale.parse(Locale('EN').toLanguageTag()).languageCode, 'en',
-        reason: 'Locale.parse does standards-compliant normalization, whereas '
-            'Locale("EN") is incorrect usage of the API as per API '
-            'documentation.');
+    expect(
+      () => Locale.parse(Locale('en', '').toString()),
+      throwsException,
+      reason: 'Locale("en", "").toString() produces "en-" which is not '
+          'standards-compliant.',
+    );
+
+    // We have:
+    expect(
+      const Locale('en'),
+      isNot(new Locale('en', '')),
+      reason: 'Legacy. (The semantic difference between Locale("en") and '
+          'Locale("en", "") is not defined.)',
+    );
+    // However we also have:
+    expect(
+      Locale.parse(Locale('en', '').toLanguageTag()),
+      Locale('en'),
+      reason: 'There is no standards-compliant way for toLanguageTag() to '
+          'represent a zero-length region code.',
+    );
+    // So this class' operator== doesn't match the behaviour we expect from the
+    // normalized Unicode BCP47 Locale Identifiers perspective.
```

Comment:
This is an important discussion point.

Reply:
We attempted to remove the empty-string use case, but it tripped some tests:
https://github.com/flutter/engine/pull/6693
I think we should try again, but it will take more work to land this.